### PR TITLE
Add project entity and relations

### PR DESCRIPTION
**Pull request by rmnbrd**

### CODE REVIEW

File: src/api/Project/mutations/createProject/index.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  _: any,
+  args: CreateProjectArgs
+): Promise<CreateProjectResponse> => {
+  const { errors, project } = await ProjectModel.createProject(args)
+
+  if (errors.length > 0) {
+    return { errors }
+  }
+
+  return { project, errors }
```
Comment (Tom-Bonnike) [inline src/api/Project/mutations/createProject/index.js]: After discussing with Paul yesterday:
For GraphQL we’ll actually do:

```js
if (errors.length > 0) {
  return { status: 'FAILURE', errors }
}

return { status: 'SUCCESS', project }
```

But we’ll do that in another PR :)

File: src/server/database/migrations/20180423233147_create-projects_users.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+export const up = async knex => {
+  await knex.schema.createTable('projects_users', table => {
```
Comment (Tom-Bonnike) [inline src/server/database/migrations/20180423233147_create-projects_users.js]: should we name the database `projectsUsers`?

File: src/models/Project/createProject/index.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// @flow
+import type { Project } from '@models/Project'
+import type { ValidationError } from '@types/ValidationError'
+import database from '@server/database'
+import ProjectUserModel from '@models/ProjectUser'
+
+export type CreateProjectArgs = {
```
Comment (PauuloG) [inline src/models/Project/createProject/index.js]: Do we all agree then that we use naming convention `ActionArgs` and `ActionResponse` for those types ?

File: src/api/Project/mutations/createProject/index.js (hunk 1 above)
Comment (rmnbrd) [reply src/api/Project/mutations/createProject/index.js]: Yes, good idea 👍

Comment (Tom-Bonnike) [reply src/api/Project/mutations/createProject/index.js]: Or we could maybe do:

```js
if (errors.length > 0) {
  throw new Error('…') // Error only takes a string…
}
```

So I think we’d have to use `JSON.stringify(errors)` an then use `apollo-server`’s `formatError` to make it an array again… That sucks.

File: src/models/ProjectUser/index.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// @flow
+import createProjectUser from './createProjectUser'
+export * from './createProjectUser'
+
+export type ProjectUser = {
+  id: number,
+  projectId: number,
+  userId: number,
+  role: string
```
Comment (Tom-Bonnike) [inline src/models/ProjectUser/index.js]: You could do: `role: ['OWNER', 'CONTRIBUTOR']` instead (we can change the actual values later)

File: src/api/schema/types/ProjectUser.graphql [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+type CreateProjectUserResponse {
+  projectUser: ProjectUser
+  errors: [ValidationError]!
+}
+
+extend type Mutation {
+  createProjectUser(
+    userId: Int!
+    projectId: Int!
+    role: String!
```
Comment (Tom-Bonnike) [inline src/api/schema/types/ProjectUser.graphql]: We could use a `GraphQLEnum` to validate the `role` (OWNER, CONTRIBUTOR to start with)

File: src/models/Project/createProject/index.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  const savedProject: Array<Project> = await database('projects')
+    .insert({ name: trimmedName, slug, description })
+    .returning('*')
+
+  await ProjectUserModel.createProjectUser({
+    projectId: savedProject[0].id,
+    userId,
+    role: 'owner'
```
Comment (Tom-Bonnike) [inline src/models/Project/createProject/index.js]: Let’s capitalize this: `OWNER`

File: src/api/schema/types/ProjectUser.graphql [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+type CreateProjectUserResponse {
+  projectUser: ProjectUser
+  errors: [ValidationError]!
+}
+
+extend type Mutation {
+  createProjectUser(
+    userId: Int!
+    projectId: Int!
```
Comment (Tom-Bonnike) [inline src/api/schema/types/ProjectUser.graphql]: Let’s use `ID!` for all user/project IDs.

File: src/models/Project/createProject/index.js (hunk 3 above)
Comment (Tom-Bonnike) [reply src/models/Project/createProject/index.js]: Or we switch to `ActionInput` and `ActionPayload`?

File: src/server/database/migrations/20180423233147_create-projects_users.js (hunk 2 above)
Comment (rmnbrd) [reply src/server/database/migrations/20180423233147_create-projects_users.js]: @PauuloG what's your opinion on this? I think we should go full camelCase 👍

File: src/models/Project/index.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+// @flow
+import createProject from './createProject'
+export * from './createProject'
+
+export type Project = {
+  id: number,
```
Comment (Tom-Bonnike) [inline src/models/Project/index.js]: I think GraphQL returns `id` as `string` and not `number`, even if you pass it a number.

Comment (rmnbrd) [reply src/models/Project/index.js]: Also fixed with 48a68a0

File: src/api/Project/mutations/createProject/integration.spec.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    await resetDatabase()
+    await database.migrate.latest()
+  })
+
+  it('should return the created project on success, with normalized and trimmed values', async done => {
+    await database.seed.run()
+    const projectResponse = await testClient.mutate({
+      variables: {
+        name: 'My awesome project',
+        slug: 'my-awesome-project',
```
Comment (Tom-Bonnike) [inline src/api/Project/mutations/createProject/integration.spec.js]: We should generate the slug automatically using: https://github.com/sindresorhus/slugify

File: src/api/Project/mutations/createProject/integration.spec.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      mutation
+    })
+    const { project, errors } = projectResponse.data.createProject
+    const expected = {
+      __typename: 'Project',
+      name: 'My awesome project'
+    }
+
+    expect(errors.length).toEqual(0)
+    expect(project).toMatchObject(expected)
```
Comment (Tom-Bonnike) [inline src/api/Project/mutations/createProject/integration.spec.js]: We could use `toEqual` by adding the other fields.

File: src/api/Project/mutations/createProject/integration.spec.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        name: 'My awesome project',
+        slug: 'my-awesome-project',
+        userId: 1
+      },
+      mutation
+    })
+    const { project, errors } = projectResponse.data.createProject
+    const expected = {
+      __typename: 'Project',
+      name: 'My awesome project'
```
Comment (Tom-Bonnike) [inline src/api/Project/mutations/createProject/integration.spec.js]: We should expect more fields. (`slug`, `description`, `id` …)

File: src/api/User/mutations/createUser/integration.spec.js [hunk 12]
```diff
@@ -1,8 +1,9 @@
 import gql from 'graphql-tag'
 import testClient from '@tests/client'
 import database from '@server/database'
+import resetDatabase from '@tests/resetDatabase'
 
-const mutation = gql`
+export const mutation = gql`
```
Comment (Tom-Bonnike) [inline src/api/User/mutations/createUser/integration.spec.js]: We don’t need to export this mutation.

File: src/models/Project/createProject/index.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  project?: Project,
+  errors: Array<ValidationError>
+}
+
+const createProject = async ({
+  name,
+  slug,
+  description,
+  userId
+}: CreateProjectArgs): Promise<CreateProjectResponse> => {
```
Comment (Tom-Bonnike) [inline src/models/Project/createProject/index.js]: We’ll need to verify auth to this method. cc @PauuloG

File: src/models/Project/createProject/index.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+  const errors: Array<ValidationError> = []
+  const trimmedName = name.trim()
+
+  if (errors.length) {
+    return { errors }
+  }
+
+  const savedProject: Array<Project> = await database('projects')
+    .insert({ name: trimmedName, slug, description })
```
Comment (Tom-Bonnike) [inline src/models/Project/createProject/index.js]: We probably should trim the description, too.

File: src/models/Project/index.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+// @flow
+import createProject from './createProject'
+export * from './createProject'
+
+export type Project = {
+  id: string,
+  name: string,
+  slug: string
```
Comment (Tom-Bonnike) [inline src/models/Project/index.js]: Description is missing.

File: src/models/ProjectUser/createProjectUser/index.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  userId,
+  role
+}: CreateProjectUserArgs): Promise<CreateProjectUserResponse> => {
+  const errors: Array<ValidationError> = []
+
+  if (errors.length) {
+    return { errors }
+  }
+
+  const savedProject: Array<ProjectUser> = await database('projectsUsers')
```
Comment (Tom-Bonnike) [inline src/models/ProjectUser/createProjectUser/index.js]: Is this the correct pluralization for a relation table? I’m not sure? cc @PauuloG

File: src/models/User/createUser/integration.spec.js [hunk 17]
```diff
@@ -89,21 +81,28 @@ describe('The User model’s `createUser` helper', () => {
...
+      passwordPlain: '$uper$trongPa$$word'
+    })
     const countAfter = await database('users').count()
 
-    expect(countBefore[0].count).toEqual('0')
-    expect(countAfter[0].count).toEqual('1')
+    expect(countBefore[0].count).toEqual('1')
+    expect(countAfter[0].count).toEqual('2')
     expect(user).toMatchObject({
-      id: 1,
```
Comment (Tom-Bonnike) [inline src/models/User/createUser/integration.spec.js]: Let’s re-add the `id` since we run in band.
I wonder, why do we not use `.toEqual` instead of `toMatchObject` here too? I forgot why I did that.

File: src/server/database/migrations/20180423095250_create-projects.js [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+  await knex.schema.createTable('projects', table => {
+    table
+      .increments('id')
+      .unsigned()
+      .primary()
+    table.string('name', 30).notNull()
+    table
+      .string('slug')
+      .unique()
+      .notNull()
```
Comment (Tom-Bonnike) [inline src/server/database/migrations/20180423095250_create-projects.js]: Slug doesn’t really have to be unique… it has to be unique for a specific user :/ I’m not sure how to deal with this

File: src/server/database/migrations/20180423095250_create-projects.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    table.text('description')
```
Comment (Tom-Bonnike) [inline src/server/database/migrations/20180423095250_create-projects.js]: Should this simply be a `string('description')`? 255 characters is more than enough.

File: src/server/database/migrations/20180423095250_create-projects.js (hunk 18 above)
Comment (Tom-Bonnike) [reply src/server/database/migrations/20180423095250_create-projects.js]: I guess the name should be unique for a specific user.

File: src/models/Project/createProject/index.js [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    .returning('*')
+
+  await ProjectUserModel.createProjectUser({
+    projectId: savedProject[0].id,
+    userId,
+    role: 'OWNER'
+  })
+
+  return { project: savedProject[0], errors: [] }
```
Comment (Tom-Bonnike) [inline src/models/Project/createProject/index.js]: I think we’re missing some edge cases. If the name is not unique, if the project doesn’t exist, … and add tests for that

File: src/models/Project/createProject/index.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+  slug,
+  description,
+  userId
+}: CreateProjectArgs): Promise<CreateProjectResponse> => {
+  const errors: Array<ValidationError> = []
+  const trimmedName = name.trim()
+
+  if (errors.length) {
+    return { errors }
+  }
```
Comment (Tom-Bonnike) [inline src/models/Project/createProject/index.js]: I know you copy-pasted but there’s no way there will be errors at this point ;p But we’re missing some validations. Unique name, project not existing, …

File: src/models/Project/createProject/index.js (hunk 14 above)
Comment (rmnbrd) [reply src/models/Project/createProject/index.js]: 👍

**Comment by Tom-Bonnike** [comment]
We have quite a few things to do before we can merge, even though this PR is already huge:
screen shot 2018-04-29 at 21 24 21

File: src/models/Project/createProject/index.js (hunk 20 above)
Comment (rmnbrd) [reply src/models/Project/createProject/index.js]: We'll do that in another PR 😊
